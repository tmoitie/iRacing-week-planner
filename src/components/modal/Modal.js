// @flow

import * as React from 'react';
import { useTranslation } from 'react-i18next';
import RemoveIcon from '../icon/RemoveIcon';

import BaseModal from './BaseModal';

import './styles/modal.scss';

import styles from '../../styles/main.module.scss';

type Props = {
  children: React.node,
  title: string,
  isOpen?: boolean,
  onClose?: () => void,
  doneButtonText?: string,
  showFooter?: boolean,
  doneAction?: () => void,
  id: string,
};

const defaultProps = {
  isOpen: false,
  onClose: () => {},
  doneAction: () => {},
  doneButtonText: 'Close',
  showFooter: true,
};

export default function Modal({
  children,
  title,
  isOpen,
  onClose,
  doneButtonText,
  showFooter,
  doneAction,
  id,
}: Props): React.Node {
  const { t } = useTranslation();

  const close = (e) => {
    e.preventDefault();
    e.stopPropagation();

    onClose(e);
  };

  const clickDone = (e) => {
    e.preventDefault();
    e.stopPropagation();

    doneAction();
  };

  React.useEffect(() => {
    const escapeModal = (e) => {
      if (e.code !== undefined && e.code === 'Escape') {
        onClose();
        e.preventDefault();
        return;
      }
      if (e.keyCode !== undefined && e.keyCode === 27) {
        onClose();
        e.preventDefault();
      }
    };

    document.addEventListener('keypress', escapeModal);

    return () => {
      document.removeEventListener('keypress', escapeModal);
    };
  }, []);

  return (
    <BaseModal isOpen={isOpen} onRequestClose={onClose} id={id}>
      <div className={styles['modal-content']}>
        <div className={styles['modal-header']}>
          <button type="button" className={styles.close} onClick={close} aria-label="Close">
            <RemoveIcon />
          </button>
          <h4 className={styles['modal-title']}>{title}</h4>
        </div>
        <div className={styles['modal-body']} style={{ maxHeight: '55vh', overflowY: 'auto' }}>
          {children}
        </div>
        {showFooter && (
          <div className={styles['modal-footer']}>
            <button type="button" className={`${styles.btn} ${styles['btn-primary']}`} onClick={clickDone}>
              {t(doneButtonText)}
            </button>
          </div>
        )}
      </div>
    </BaseModal>
  );
}

Modal.defaultProps = defaultProps;
