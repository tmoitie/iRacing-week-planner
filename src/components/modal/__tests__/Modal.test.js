import { describe, test } from '@jest/globals';
import React from 'react';
import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Modal from '../Modal';

describe('components/modal/Modal', () => {
  test('renders closed', async () => {
    const { baseElement } = render(
      <Modal isOpen={false} onClose={() => {}} title="Test Closed" id="testClosed">
        <div>Hello Closed!</div>
      </Modal>,
    );

    expect(screen.queryByTestId('testClosed')).not.toBeTruthy();
  });

  test('renders no footer', async () => {
    const { baseElement } = render(
      <Modal isOpen onClose={() => {}} title="Test No Footer" showFooter={false} id="testNoFooter">
        <div>No Footer!</div>
      </Modal>,
    );

    expect(await screen.findByTestId('testNoFooter')).toMatchSnapshot();
  });

  test('renders open', async () => {
    const user = userEvent.setup();
    const onClose = jest.fn(() => {});
    const doneAction = jest.fn(() => {});

    const { baseElement } = render(
      <Modal
        isOpen
        onClose={onClose}
        title="Test Open"
        doneButtonText="Is Done"
        doneAction={doneAction}
        id="testOpen"
      >
        <div>Hello Open!</div>
      </Modal>,
    );

    expect(await screen.findByTestId('testOpen')).toMatchSnapshot();

    await user.click(screen.getByLabelText('Close'));
    expect(onClose).toHaveBeenCalled();

    await user.click(screen.getByText('Is Done'));
    expect(doneAction).toHaveBeenCalled();
  });

  test('presses escape', async () => {
    const onClose = jest.fn(() => {});
    const { baseElement, unmount } = render(
      <Modal isOpen onClose={onClose} title="Test Escape" />,
    );

    document.dispatchEvent(new KeyboardEvent('keypress', { code: 'KeyR', key: 82, keyCode: 82 }));
    expect(onClose).not.toHaveBeenCalled();

    document.dispatchEvent(new KeyboardEvent('keypress', { code: 'Escape', key: 27, keyCode: 27 }));
    expect(onClose).toHaveBeenCalledTimes(1);

    // Backwards compatibility
    document.dispatchEvent(new KeyboardEvent('keypress', { keyCode: 27 }));
    expect(onClose).toHaveBeenCalledTimes(2);

    unmount();

    document.dispatchEvent(new KeyboardEvent('keypress', { code: 'Escape', key: 27, keyCode: 27 }));
    expect(onClose).toHaveBeenCalledTimes(2);
  });

  test('default actions don\'t crash', async () => {
    const user = userEvent.setup();

    render(<Modal isOpen title="Test Open" doneButtonText="OK" />);

    await user.click(screen.getByLabelText('Close'));
    await user.click(screen.getByText('OK'));
  });
});
