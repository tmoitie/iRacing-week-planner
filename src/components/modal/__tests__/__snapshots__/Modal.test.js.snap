// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/Modal renders no footer 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="testNoFooter"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Test No Footer
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div>
        No Footer!
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/Modal renders open 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="testOpen"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Test Open
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div>
        Hello Open!
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Is Done
      </button>
    </div>
  </div>
</div>
`;
